composer/models/nlp_metrics.py

```diff
@@ -3,12 +3,159 @@
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+
+        self.vocab_size = vocab_size
+        self.ignore_index = ignore_index
+        self.loss_fn = torch.nn.CrossEntropyLoss(ignore_index=ignore_index, reduction="sum")
+        self.add_state("sum_loss", default=torch.tensor(0.), dist_reduce_fx="sum")
+        self.add_state("total_batches", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, output: Union[Mapping, Tensor], target: Tensor) -> None:
+        """Updates the internal state with results from a new batch.
+
+        Args:
+            output (Mapping): The output from the model, which must contain
+                either the Tensor or a Mapping type that contains the loss or model logits.
+            target (Tensor): A Tensor of ground-truth values to compare against.
+        """
+
+        assert isinstance(output, Tensor)
+        output = output.view(-1, self.vocab_size)
+        target = target.view(-1)
+        losses = self.loss_fn(output, target)
+
+        if self.ignore_index:
+            mask = target != self.ignore_index
+            total_batches = mask.sum()
+        else:
+            total_batches = target.shape[0]
+
+        self.total_batches += total_batches  #type: ignore (third-party)
+
+        # accmulate loss over all batches
+        self.sum_loss += losses
+
+    def compute(self) -> Tensor:
+        """Aggregate the state over all processes to compute the metric.
+
+        Returns:
+            loss (Tensor): The loss averaged across all batches.
+        """
+        # Return average loss over entire dataset
+        return self.sum_loss / self.total_batches  #type: ignore (third-party)
+
+
+# TODO (Moin): write tests for this!
+class BinaryF1Score(Metric):
+    """Implements F1 Scores for binary classification tasks via sklearn.
+
+    Args:
+        dist_sync_on_step (bool): Synchronize metric state across processes at
+            each forward() before returning the value at the step.
+
+    State:
+        sum_loss (float): the sum of the per-example loss in the batch.
+        total_batches (float): the number of batches to average across.
+    """
+
+    def __init__(self, dist_sync_on_step=False):
+        super().__init__(dist_sync_on_step=dist_sync_on_step)
+
+        self.add_state("predictions", default=[], dist_reduce_fx="cat")
+        self.add_state("labels", default=[], dist_reduce_fx="cat")
+
+    def update(self, output: Union[Mapping, Tensor], target: Tensor) -> None:
+        """Updates the internal state with results from a new batch.
+
+        Args:
+            output (Mapping): The output from the model, which must contain
+                either the Tensor or a Mapping type that contains the loss or model logits.
+            target (Tensor): A Tensor of ground-truth values to compare against.
+        """
+        assert isinstance(self.predictions, list)
+        assert isinstance(self.labels, list)
+        self.predictions.append(output)
```

Comment:
TBH, I think it would cleaner and fairly easy to calculate the true positive (tp), false positive (fp), and false negative (fn) rates in the update, then manually compute the F1 score tp / (tp + 0.5 * (fp + fn)) from [here](https://en.wikipedia.org/wiki/F-score)

```suggestion
        prediction = output.argmax(dim=1)
        self.tp += prediction[target].sum()
        self.fp += (prediction[target] == 0).sum()
        self.fn += (prediction[target == 0] == 1).sum()
```

Up to your discretion since this is already working 🙂

Reply:
The more we talk about metrics, the more I realize that we should fix it. I'll implement your suggestion + add unit tests against sklearn's implementation today!

Reply:
@Landanjs I just adopted your suggestion and ensured that it properly works!